File: model-client/src/main/java/org/modelix/model/persistent/CPHamtNode.java
package org.modelix.model.persistent;

import java.util.Arrays;
import java.util.function.Function;

public abstract class CPHamtNode {
  public static final Function<String, CPHamtNode> DESERIALIZER = s -> deserialize(s);
  public static CPHamtNode deserialize(String input) {
    String[] parts = input.split("/", -1);

    if ("L".equals(parts[0])) {
      return new CPHamtLeaf(SerializationUtil.longFromHex(parts[1]), parts[2]);
    } else if ("I".equals(parts[0])) {
      return new CPHamtInternal(
              SerializationUtil.intFromHex(parts[1]),
              Arrays.stream(parts[2].split(","))
                      .filter(it -> (it != null && it.length() > 0))
                      .toArray(String[]::new));
    } else {
      throw new RuntimeException("Unknown type: " + parts[0] + ", input: " + input);
    }
  }

  public abstract String serialize();
}

File: model-client/src/main/java/org/modelix/model/persistent/CPHamtNode.kt
package org.modelix.model.persistent

import org.modelix.model.persistent.SerializationUtil.intFromHex
import org.modelix.model.persistent.SerializationUtil.longFromHex
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

abstract class CPHamtNode {
    abstract fun serialize(): String?

    companion object {
        val DESERIALIZER = Function { s: String? -> deserialize(s!!) }
        @JvmStatic
        fun deserialize(input: String): CPHamtNode {
            val parts = input.split("/").dropLastWhile { it.isEmpty() }.toTypedArray()
            return if ("L" == parts[0]) {
                CPHamtLeaf(longFromHex(parts[1]), parts[2])
            } else if ("I" == parts[0]) {
                CPHamtInternal(
                        intFromHex(parts[1]),
                        Arrays.stream(parts[2].split(",").toTypedArray())
                                .filter { it: String? -> it != null && it.length > 0 }
                                .collect(Collectors.toList()).toTypedArray())
            } else {
                throw RuntimeException("Unknown type: " + parts[0] + ", input: " + input)
            }
        }
    }
}